---
title: "DiversityV1"
output: html_document
date: "2022-11-28"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
install.packages("vegan")
install.packages("ape")
install.packages("phangorn")
install.packages("seqinr")
install.packages("cowplot")
```

Had to manually install dependency of MicrobiomeMarker due to issue with repository
```{r}
install.packages("~/Downloads/densvis_1.10.1.tgz", repos = NULL, type = .Platform$pkgType)
```

You should already have this one
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.17")
```

```{r}
BiocManager::install("phyloseq")
```

```{r}
BiocManager::install("microbiome")
BiocManager::install("ALDEx2")
BiocManager::install("microbiomeMarker")
BiocManager::install("dada2")
BiocManager::install("DEFormats")
```

```{r}
library(phyloseq)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(phyloseq)
library(microbiome)
library(microbiomeMarker)
library(ALDEx2)
library(vegan)
library(cowplot)
library(lefser)
library(DEFormats)
library(phyloseq)
library(DESeq2)
library(purrr)
```

Load the files and data

LOAD ENVIRONMENT
```{r}
load("CT.RData")
```

LOAD METADATA
```{r}
metadata<-read.delim("CTmetadata.txt")
metadata$sample.names
```
```{r}
data.frame(colnames(t(seqtab.nochim)), metadata$sample.names)
```
```{r}
rownames(metadata) <- rownames(seqtab.nochim)
metadata$sample.names<-rownames(metadata)
```

COMPLETE
```{r}
seqs <- getSequences(seqtab.nochim)
names(seqs) <- seqs # This propagates to the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA)

```

COMPLETE
```{r}

phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phang.align)
treeNJ<- NJ(dm)
```

COMPLETE
```{r}
fit = pml(treeNJ, data=phang.align)

fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                      rearrangement = "stochastic", control = pml.control(trace = 0))

```

COMPLETE
```{r}
dna <- DNAStringSet(getSequences(seqtab.nochim)) # Create a DNAStringSet from the ASVs
ids <- IdTaxa(dna, trainingSet, strand="top", processors=NULL, verbose=FALSE) # use all processors

ranks <- c("domain", "phylum", "class", "order", "family", "genus", "species") # ranks of interest
# Convert the output object of class "Taxa" to a matrix analogous to the output from assignTaxonomy
taxid <- t(sapply(ids, function(x) {
        m <- match(ranks, x$rank)
        taxa <- x$taxon[m]
        taxa[startsWith(taxa, "unclassified_")] <- NA
        taxa
}))
colnames(taxid) <- ranks; rownames(taxid) <- getSequences(seqtab.nochim)
```

CAN RUN FROM HERE
```{r}
colnames(taxid)<-c("Kingdom",  "Phylum" , "Class" ,  "Order" ,  "Family",  "Genus"  , "Species")
```

Create a phyloseq object
```{r}
ps <- phyloseq(tax_table(taxid), sample_data(metadata),
                 otu_table(seqtab.nochim,taxa_are_rows=FALSE, phy_tree(fitGTR$tree)))
```

Changes the ASV names
```{r}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

# Alpha diversity

```{r}
rare_data<-rarecurve(seqtab.nochim, step = 20, sample=5000, col = "blue", cex = 0.6, tidy=TRUE)
```

```{r}
rare_data$sample.names<-rare_data$Site
```

```{r}
rare_data<-rare_data%>%left_join(metadata)
```

```{r}
rare_data%>%ggplot(aes(Sample, Species, group=Site))+geom_line()+geom_vline(xintercept=40000)+theme_bw()
ggsave("fig1.png")
ggsave("fig1.pdf")
```
```{r}
reads<-ggplot(data = data.frame("total_reads" =  phyloseq::sample_sums(ps),
                         "observed" = phyloseq::estimate_richness(ps, measures = "Observed")[, 1]),
       aes(x = total_reads, y = observed)) +
  geom_point() +
  geom_smooth(method="lm", se = FALSE) +
  labs(x = "\nTotal Reads", y = "Observed Richness\n")
reads+theme_bw()
```

Normalise data 
```{r}
ps_rare <- phyloseq::rarefy_even_depth(ps,sample.size=40000, rngseed = 123, replace = FALSE)
```

Generate the alpha diversity
```{r}
ps_alpha_rich <- estimate_richness(ps_rare, split = TRUE)
```

```{r}
ps_alpha_rich$sample.names<-rownames(ps_alpha_rich)
```

Join the alpha diversity with the metadata
```{r}
ps_alpha_rich<-ps_alpha_rich%>%left_join(metadata)
```

Plot the data
```{r}
a<-ps_alpha_rich%>%ggplot(aes(Group, Shannon ))+geom_boxplot()+geom_jitter()+theme_bw()
b<-ps_alpha_rich%>%ggplot(aes(Group, Chao1))+geom_boxplot()+geom_jitter()+theme_bw()
plot_grid(a,b)
ggsave("fig2.png")
ggsave("fig2.pdf")
```
```{r}
ps_alpha_rich$GENOTYPE<-factor(ps_alpha_rich$GENOTYPE)
ps_alpha_rich$TIME<-factor(ps_alpha_rich$TIME)
a<-ps_alpha_rich%>%ggplot(aes(TIME, Shannon, group=MOUSE_code))+geom_line()+facet_grid(.~GENOTYPE)+theme_bw()
b<-ps_alpha_rich%>%ggplot(aes(TIME, Chao1, group=MOUSE_code))+geom_line()+facet_grid(.~GENOTYPE)+theme_bw()
plot_grid(a,b)
ggsave("fig3.png")
ggsave("fig3.pdf")
```
make a summary

```{r}
ave_alpha<-ps_alpha_rich%>%filter(Group!="AKR1D1")%>%group_by(Group)%>%summarise(mean_shannon=mean(Shannon), sd_shannon=sd(Shannon))
```

Plot Alpha diversity
```{r}
alpha_bar<-ave_alpha%>%ggplot(aes(Group, mean_shannon, fill=Group))+geom_bar(stat="identity", position = position_dodge(),colour="black")+
geom_errorbar(aes(ymin = mean_shannon-sd_shannon, ymax = mean_shannon+sd_shannon),position=position_dodge(.9), width=0.6)+
scale_y_continuous(limits = c(0, 10), expand = c(0, 0))+
theme_bw()+
ylab("mean Shannon diversity")
alpha_bar

ggsave("fig4.png")
ggsave("fig4.pdf")
```

Statistics

Non parametric tests
```{r}
kruskal.test(ps_alpha_rich$Shannon, ps_alpha_rich$Group)
pairwise.wilcox.test(ps_alpha_rich$Shannon, ps_alpha_rich$Group, p.adjust.method = "BH")
```

# Beta diversity

```{r}
pslog <- transform_sample_counts(ps_rare, function(x) log(1 + x))
out.bc.log <- ordinate(pslog, method = "MDS", distance = "bray")
out.jac.log <- ordinate(pslog, method = "MDS", distance = "jaccard")
```

```{r}
evals <- out.bc.log$values$Eigenvalues
p <- plot_ordination(pslog, out.bc.log) +
  coord_fixed(sqrt(evals[2] / evals[1])) +
  theme_bw() +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point(aes(color = Group, x = Axis.1, y = Axis.2)) +
  scale_color_manual(values = c("CRE0" = "cyan", "CRE12" = "blue","WT0" = "hotpink","WT12" = "red"))

print(p)
ggsave("fig5.png")
ggsave("fig5.pdf")
```

```{r}
pseq.rel <- microbiome::transform(ps_rare, "compositional")
otu <- microbiome::abundances(pseq.rel)
meta <- microbiome::meta(pseq.rel)
```

Permanova for GENOTYPE
```{r}
# Compute the distance matrices for Jaccard and Bray-Curtis distances
jaccard_dist <- vegdist(t(otu_table(pslog)), method = "jaccard")
bray_dist <- vegdist(t(otu_table(pslog)), method = "bray")

# Create a data frame with the sample data
sample_data_df <- data.frame(sample_data(pslog))

# Perform PERMANOVA tests
braypermanova_GENOTYPE <- adonis2(t(otu) ~ GENOTYPE, data = meta, permutations=999, method = "bray")
print(braypermanova_GENOTYPE)
jacpermanova_GENOTYPE <- adonis2(t(otu) ~ GENOTYPE, data = meta, permutations=999, method = "jaccard")
print(jacpermanova_GENOTYPE)

# Check that variance homogeneity assumptions hold
dist <- vegdist(t(otu),"bray")
anova(betadisper(dist, meta$GENOTYPE))

dist <- vegdist(t(otu),"jaccard")
anova(betadisper(dist, meta$GENOTYPE))
```

Permanova for Groups:
```{r}
braypermanova <- adonis2(t(otu) ~ Group, data = meta, permutations=999, method = "bray")
print(braypermanova)
jacpermanova <- adonis2(t(otu) ~ Group, data = meta, permutations=999, method = "jaccard")
print(jacpermanova)
```

Check that variance homogeneity assumptions hold
```{r}
print("bray permanova")
dist <- vegdist(t(otu),"bray")
anova(betadisper(dist, meta$Group))

print("jaccard permanova")
dist <- vegdist(t(otu),"jaccard")
anova(betadisper(dist, meta$Group))
```

```{r}
dispr <- vegan::betadisper(dist, phyloseq::sample_data(ps_rare)$Group)
dispr
boxplot(dispr, main = "", xlab = "")

ggsave("fig6.png")
ggsave("fig6.pdf")
```

# TAXONOMY

Prepare taxonomic data
```{r}
rel_ab<-t(otu_table(ps_rare))
rel_ab<-apply(rel_ab, 2, function(x) x/sum(x)*100)
rel_ab<-data.frame(rel_ab)
rel_ab$ID<-rownames(rel_ab)
taxa<-data.frame(tax_table(ps_rare))
taxa$ID<-rownames(taxa)
rel_ab<-data.frame(rel_ab%>%left_join(taxa))
```

Create taxa bar plot
```{r}
d<-pivot_longer(rel_ab,starts_with("CT"), values_to="Ab", names_to="sample.names")
d<-d%>%group_by(Order,sample.names)%>%summarise(per=sum(Ab))
d<-d%>%left_join(metadata)
d<-d%>%group_by(Order,Group)%>%summarise(ave.per=mean(per))
d<-d%>%mutate(Order=replace(Order, ave.per<1, "Other")) %>% as.data.frame()

d$Phyum<-factor(d$Order)
taxa_plot<-d%>%ggplot(aes(x=Group, y=ave.per, fill=Order)) + geom_bar(stat="identity")+theme_bw()+scale_y_continuous(limits = c(0, 104), expand = c(0,0))+theme(legend.position="right")
taxa_plot
ggsave("fig7b.png")
ggsave("fig7b.pdf")
print(d)
```

taxa_plot+theme(legend.position="right")
```{r}
ps_order <- phyloseq::tax_glom(ps_rare, "Order")
phyloseq::taxa_names(ps_order) <- phyloseq::tax_table(ps_order)[, "Order"]
order_plots<-phyloseq::psmelt(ps_order)
```

Add colours as scale_color_manual
```{r}
order_plots_filtered <- order_plots %>%
  filter(Order %in% c("Lachnospirales", "Bifidobacteriales", "Bacteroidales"))

fig <- order_plots_filtered %>%
  ggplot(aes(Group, Abundance, fill=Group)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(alpha=0.3, size=1) +
  labs(x = " ", y = "Abundance\n") +
  facet_wrap(~ OTU, scales = "free")

print(fig)

ggsave("fig8.png")
ggsave("fig8.pdf")
```

Abundance test by Order
```{r}
order_plots_filtered$OTU<-factor(order_plots_filtered$OTU)
levels(order_plots_filtered$OTU)
order_plots_filtered%>%split(.$OTU) %>%map(~ kruskal.test(Abundance ~ GENOTYPE, data = .x)) 
```

preparing data for log fold change
```{r}
# Convert phyloseq object to DESeq2 object
dds <- phyloseq_to_deseq2(ps, ~ GENOTYPE)

# Run DESeq2
dds <- DESeq(dds)

# Create contrasts
con1 <- c("GENOTYPE", "WT", "CRE")

# Perform differential gene expression analysis for each contrast
res1 <- results(dds, contrast = con1, alpha = 0.1)
```

Convert results to data frames and add taxa information
```{r}
DESeq2_res_df1 <- data.frame(res1)
DESeq2_res_df1$taxa <- rownames(DESeq2_res_df1)

taxa <- data.frame(tax_table(ps_rare))
taxa$taxa <- rownames(taxa)

results1_DESeq2 <- taxa %>% left_join(DESeq2_res_df1, by = "taxa")
```

Filter the results based on FDR (padj < 0.05)
```{r}
fdr_DESeq21 <- results1_DESeq2 %>%
    dplyr::filter(padj < 0.05)
```

log fold change plots (by Order)
```{r}
# Create the plot for the first contrast (WT vs. CRE)
vol1 <- fdr_DESeq21 %>%
  ggplot(aes(x = Order, y = log2FoldChange)) +
  geom_point(size = 4) +
  labs(y = "\nLog2 Fold-Change for WT vs. CRE", x = "") +
  theme(axis.text.x = element_text(color = "black", size = 12),
        axis.text.y = element_text(color = "black", size = 12),
        axis.title.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 12),
        legend.position = "none") +
  coord_flip() +
  geom_hline(yintercept = 0, linetype = "dotted") + theme_bw()
print(vol1)

ggsave("fig9.png")
ggsave("fig9.pdf")


```

Print log fold change values and taxa for WT vs. CRE
```{r}
cat("Log2 Fold-Change for WT vs. CRE:\n")
cat(paste(fdr_DESeq21$taxa, fdr_DESeq21$Order, fdr_DESeq21$log2FoldChange, sep = ": "), sep = "\n")
cat(paste(fdr_DESeq21$taxa, "padj", fdr_DESeq21$padj, sep = ": "), sep = "\n")
cat("\n\n")
```

# LEfSe

```{r}
#gradually lowered cutoffs to reveal the largest factors 
#did not include TIME results in paper due to non-significance
#TIME
mm_lefse <- run_lefse(
    ps_rare,
    wilcoxon_cutoff = 0.1,
    group = "TIME",
    kw_cutoff = 0.1,
   taxa_rank = "Order",
    multigrp_strat = TRUE,
    lda_cutoff = 2)
plot_ef_bar(mm_lefse, label_level = 2,max_label_len = 100)

#GENOTYPE
mg_lefse <- run_lefse(
    ps_rare,
    wilcoxon_cutoff = 0.05,
    group = "GENOTYPE",
    kw_cutoff = 0.1,
   taxa_rank = "Order",
    multigrp_strat = TRUE,
    lda_cutoff = 4)
plot_ef_bar(mg_lefse, label_level = 2, max_label_len = 100)

ggsave("fig10.png")
ggsave("fig10.pdf")

```